### PR TITLE
ENH: Added index to testing assert message when series values differ

### PR DESCRIPTION
**Pull request by amilbourne**
- [ ] closes #xxxx
- [ ] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [ ] whatsnew entry

This implements the suggestion made in #31190 
In short, where the values in 2 series differ, the assert exception message includes the index as well as the values.  This helps when using assert_frame_equal with check_like=True, since the index may be reordered, making the output hard to understand.

### CODE REVIEW

File: pandas/_testing.py [hunk 1]
```diff
@@ -956,6 +965,8 @@ def assert_numpy_array_equal(
     obj : str, default 'numpy array'
         Specify object name being compared, internally used to show appropriate
         assertion message.
+    index : numpy.ndarray
```
Comment (MarcoGorelli) [inline pandas/_testing.py]: Could you specify the default value (None) here?

**Review by MarcoGorelli — CHANGES_REQUESTED** [review]

File: pandas/_testing.py (hunk 1 above)
Comment (amilbourne) [reply pandas/_testing.py]: Yes - I missed that.  Will fix.

File: pandas/tests/util/test_assert_series_equal.py [hunk 2]
```diff
@@ -162,12 +162,14 @@ def test_series_equal_length_mismatch(check_less_precise):
 
     with pytest.raises(AssertionError, match=msg):
         tm.assert_series_equal(s1, s2, check_less_precise=check_less_precise)
+        print("No Exception")
```
Comment (WillAyd) [inline pandas/tests/util/test_assert_series_equal.py]: Can remove this

**Review by WillAyd — CHANGES_REQUESTED** [review]
Reading through your issue wouldn't you want the index of both objects to be printed? If not which index gets printed here? The left?

File: pandas/tests/util/test_assert_series_equal.py (hunk 2 above)
Comment (amilbourne) [reply pandas/tests/util/test_assert_series_equal.py]: Yes. This was left over from some early debugging - and I missed it.  Will remove.

**Comment by amilbourne** [comment]
@WillAyd : 
The indexs are compared before the content, so if the indexs differ the changes in this PR never come into play.

If the user calls assert_frame_equal with check_like=False (the default) then the indexs must contain the same values in the same order, so they will be identical.  Hence no need to specify which one is being printed.

If the user calls assert_frame_equal with check_like=True, then the indexes must contain the same values, but the order may differ.  However, the field values for each index (row) must match for the frames to be considdered equal.  The (existing) code reindexes one of the frames in the order of the other and then compares them.  Presuming the indexes have the same values but were ordered differently then one of the frames will have its rows re-ordered.  This was previously confusing in the output, as there was no indication of which rows had moved or of their new order.  This PR addresses that by adding the index.  Because both frames must have the same values in their index (or the code wouldn't have got as far as checking the row values), it doesn't matter which index is being used, since the row values in each frame would have had the same index value.

I hope that makes sense; I'm not sure how well I explained it there.

**Comment by amilbourne** [comment]
Looks like the build failure was spurious.  Will hopefully vanish when I commit changes to address comments above.

**Comment by amilbourne** [comment]
@WillAyd Does the description above make sense?

**Comment by amilbourne** [comment]
@WillAyd Sorry if you wanted me to make a change, but I think I have addressed your comments.  GitHub still thinks there is a change requested though, and I can't figure out what I need to do to resolve it.

**Comment by WillAyd** [comment]
So IIUC this only applies to `check_like` because we align the index before comparing the frames, but it's not clear in case of failures what alignment was actually used, right?

**Comment by amilbourne** [comment]
Yes, this change is specifically intended to address an issue with index reordering if check_like=True.

The extra output (the index values) is also generated when check_like=False because it is equally valid and it keeps the output consistent, but when chec_like=False it doesn't add all that much, as the user would already know the index order.  That said, the index values may make the output easier to refer to if the index is unordered or hard to remember.

File: pandas/_libs/testing.pyx [hunk 3]
```diff
@@ -89,6 +89,9 @@ cpdef assert_almost_equal(a, b,
     robj : str, default None
         Specify right object name being compared, internally used to show
         appropriate assertion message
+    index : str, default None
```
Comment (jreback) [inline pandas/_libs/testing.pyx]: can you add a versionadded tag 1.1.0

File: pandas/_testing.py [hunk 4]
```diff
@@ -917,9 +924,10 @@ def raise_assert_detail(obj, message, left, right, diff=None):
     elif is_categorical_dtype(right):
         right = repr(right)
 
-    msg = f"""{obj} are different
+    if isinstance(index, np.ndarray):
```
Comment (jreback) [inline pandas/_testing.py]: what does this do?

File: pandas/_testing.py [hunk 5]
```diff
@@ -956,6 +965,8 @@ def assert_numpy_array_equal(
     obj : str, default 'numpy array'
         Specify object name being compared, internally used to show appropriate
         assertion message.
+    index : numpy.ndarray, default None
```
Comment (jreback) [inline pandas/_testing.py]: why are you adding this here? this by definition *only* compares numpy array objects

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_testing.py (hunk 5 above)
Comment (amilbourne) [reply pandas/_testing.py]: In _testing.py, assert_frame_equal compares the content of the frame one column at a time, by calling assert_series_equal (1389).  This in some (most?) cases converts the series to ndarrays and calls assert_numpy_array_equal to compare them.

File: pandas/_testing.py (hunk 4 above)
Comment (amilbourne) [reply pandas/_testing.py]: Oops - nothing.
Left over after refactor and should have been removed.  Will remove.  Thanks.

File: pandas/_libs/testing.pyx (hunk 3 above)
Comment (amilbourne) [reply pandas/_libs/testing.pyx]: Will do.

**Comment by amilbourne** [comment]
I just rebased this change against the latest in master to resolve the merge conflict and keep things clean.  Sadly I missed a Pep 8 issue because the rebase didn't invoke the commit hooks :-/

**Comment by amilbourne** [comment]
@jreback I think I have addressed all your comments but I am unsure of the etiquette here.  Should I click on the 'resolve conversation' button?  2 of your comments are probably resolved (presuming my changes were what you wanted) but 1 was a question and I am unsure whether my answer is sattisfactory.

File: pandas/_testing.py [hunk 6]
```diff
@@ -1142,7 +1150,11 @@ def assert_series_equal(
             raise AssertionError("check_exact may only be used with numeric Series")
 
         assert_numpy_array_equal(
-            left._values, right._values, check_dtype=check_dtype, obj=str(obj)
+            left._values,
+            right._values,
+            check_dtype=check_dtype,
+            obj=str(obj),
+            index=left.index._internal_get_values(),
```
Comment (jreback) [inline pandas/_testing.py]: we are really trying to remove this, just use ``np.asarray(left.index)``

File: pandas/_testing.py [hunk 7]
```diff
@@ -1181,6 +1193,7 @@ def assert_series_equal(
             check_less_precise=check_less_precise,
             check_dtype=check_dtype,
             obj=str(obj),
+            index=left.index._internal_get_values(),
```
Comment (jreback) [inline pandas/_testing.py]: same

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/_testing.py (hunk 6 above)
Comment (amilbourne) [reply pandas/_testing.py]: done

File: pandas/_testing.py (hunk 7 above)
Comment (amilbourne) [reply pandas/_testing.py]: done

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
lgtm. @WillAyd @jorisvandenbossche @jbrockmendel if any comments.

File: pandas/_libs/testing.pyx [hunk 8]
```diff
@@ -89,6 +89,12 @@ cpdef assert_almost_equal(a, b,
...
```
Comment (jbrockmendel) [inline pandas/_libs/testing.pyx]: can this be "label" or something less likely to be confused for an Index object

File: pandas/_testing.py [hunk 9]
```diff
@@ -1181,6 +1193,7 @@ def assert_series_equal(
             check_less_precise=check_less_precise,
             check_dtype=check_dtype,
             obj=str(obj),
+            index=np.asarray(left.index),
```
Comment (jbrockmendel) [inline pandas/_testing.py]: the docstring in assert_almost equal says this is a string or None

Comment (amilbourne) [reply pandas/_testing.py]: Yes it does :-(
Sorry - now fixed - thanks.

File: pandas/_libs/testing.pyx (hunk 8 above)
Comment (amilbourne) [reply pandas/_libs/testing.pyx]: I changed all the 'index' parameters to 'index_values' - I hope that is clearer.

**Comment by jreback** [comment]
thanks @amilbourne nice addition!

**Pull request merged**